cachi2/core/package_managers/npm.py

```diff
@@ -173,6 +185,199 @@ def get_sbom_components(self) -> list[dict[str, str]]:
+            info["repo"],
+            f'{info["repo"]}-external-gitcommit-{info["ref"]}.tgz',
+        )
+    )
+
+    # Create missing directories
+    directory = os.path.dirname(download_path)
+    os.makedirs(directory, exist_ok=True)
+    clone_as_tarball(info["url"], info["ref"], download_path.path)
+
+    return download_path
+
+
+def _get_npm_dependencies(
+    download_dir: RootedPath, deps_to_download: Dict[str, Dict[str, str]]
+) -> Dict[str, RootedPath]:
+    """
+    Download npm dependencies.
+
+    Receives the destination directory (download_dir)
+    and the dependencies to be downloaded (deps_to_download).
+
+    :param download_dir: Destination directory path where deps will be downloaded
+    :param deps_to_download: Dict of dependencies to be downloaded.
+    :return: Dictionary of dependencies with downloaded paths
+    """
+    files_to_download: dict[str, dict[str, Any]] = {}
+    download_paths = {}
+    for dependency, info in deps_to_download.items():
+        if (
+            info["url"].startswith("github:")
+            or info["url"].startswith("gitlab:")
+            or info["url"].startswith("bitbucket:")
+        ):
+            info["url"] = _update_vcs_url_with_full_hostname(info["url"])
+
+        if info["url"].startswith("git+"):
+            # Git source
+            download_paths[dependency] = _clone_repo_pack_archive(info["url"], download_dir)
+        else:
+            if "registry.npmjs.org" == urlparse(info["url"]).hostname:
+                # Tar archive from npm registry
+                archive_name = f'{dependency}-{info["version"]}.tgz'.removeprefix("@").replace(
+                    "/", "-"
+                )
+                download_paths[dependency] = download_dir.join_within_root(archive_name)
+            else:
+                # Tar archive
+                if info["integrity"]:
+                    algorithm, digest = ChecksumInfo.from_sri(info["integrity"])
+                else:
+                    raise PackageRejected(
+                        f"{dependency} is missing integrity checksum. It is mandatory"
+                        f"for https dependencies.",
+                        solution="Please double-check provided package-lock.json that"
+                        " your dependencies specify integrity. Try to "
+                        "rerun `npm install` on your repository.",
+                    )
+                download_paths[dependency] = RootedPath(
+                    Path(
+                        download_dir.path,
+                        f"external-{dependency}",
+                        f"{dependency}-external-{algorithm}-{digest}.tgz",
+                    )
+                )
+
+                # Create missing directories
+                directory = os.path.dirname(download_paths[dependency])
+                os.makedirs(directory, exist_ok=True)
+
+            files_to_download[info["url"]] = {
+                "download_path": download_paths[dependency],
```

Comment:
I only noticed this now -- are the keys in `download_paths` just the names of dependencies? If the lockfile has two different versions of the same dependency, what happens?

Reply:
ack It makes sense